### ISSUE TITLE
graph coloring wiki/example and unit tests hit infinite loop with Kokkos' develop and Kokkos_ENABLE_DEBUG_BOUNDS_CHECK on Pre-Volta GPUs

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos-kernels#785]
Testing for 3.2 release exposed a failing configuration in the graph coloring tests and wiki example with Kokkos' develop branch, but not Kokkos' master, when enabling `Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON` on pre-Volta NVIDIA GPUs (I confirmed the issue did not occur on a machine with Volta70 GPUs).

Kokkos core and containers unit tests with matching configuration pass; the algorithms `cuda.Random_XorShift1024` unit test consistently fails in this configuration on develop, but passes on master. The Kokkos random generator does not appear to be used in the graph coloring algorithms so I don't think this failure is relevant here. 
```bash
/ascldap/users/ndellin/kokkos-testing/algorithms/unit_tests/TestRandom.hpp:408: Failure
Value of: 1
Expected: test_int.pass_var
Which is: 0
```

Posting issue here since this occurs in the graph coloring unit tests (and wiki example), but not certain if this is a graph coloring bug, or if it is touching an issue in Kokkos not caught in core unit testing.

Reproducer instructions for the White or Ride testbeds (Power8, Pascal60 archs) follow, though this is likely reproducible on a system with a Pascal GPU.

Note the configurations include details for enabling tpls, this was a byproduct of reproducing configurations details from the failing Trilinos build that initially exposed the problem.

**Failing KokkosKernels configuration for Power8,Pascal60**
```bash
# Modules 
module load cuda/9.2.88 cmake/3.12.3 gcc/7.2.0
module load netlib/3.8.0/gcc/7.2.0

# TPL setup
BLAS_LIBRARY_DIRS="${BLAS_ROOT}/lib"
LAPACK_LIBRARY_DIRS="${BLAS_ROOT}/lib"
BLAS_LIBRARIES="blas"
LAPACK_LIBRARIES="lapack"
KOKKOSKERNELS_TPL_PATH_CMD="--user-blas-path=${BLAS_LIBRARY_DIRS} --user-lapack-path=${LAPACK_LIBRARY_DIRS}"
KOKKOSKERNELS_TPL_LIBS_CMD="--user-blas-lib=${BLAS_LIBRARIES} --user-lapack-lib=${LAPACK_LIBRARIES}"
KOKKOSKERNELS_EXTRA_LINKER_FLAGS_CMD="--extra-linker-flags=-lgfortran,-lm"

# Kokkos configure
cmake -DCMAKE_CXX_COMPILER=$KOKKOS_PATH/bin/nvcc_wrapper -DCMAKE_CXX_FLAGS= -DCMAKE_EXE_LINKER_FLAGS= -DCMAKE_INSTALL_PREFIX=$KOKKOS_INSTALL_PATH -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_PASCAL60=ON -DKokkos_ARCH_POWER8=ON -DKokkos_ENABLE_TESTS=OFF -DKokkos_ENABLE_EXAMPLES=OFF -DKokkos_ENABLE_CUDA_LAMBDA=ON -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_BUILD_TYPE=DEBUG -DKokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON -DKokkos_ENABLE_DEBUG=ON $KOKKOS_PATH

# Next, install Kokkos

# KokkosKernels configure
cmake -DCMAKE_CXX_COMPILER=$KOKKOS_PATH/bin/nvcc_wrapper -DKokkos_DIR=$KOKKOS_INSTALL_PATH/lib64/cmake/Kokkos -DCMAKE_CXX_FLAGS= -DCMAKE_INSTALL_PREFIX= -DKokkosKernels_ENABLE_TESTS=ON -DKokkosKernels_ENABLE_EXAMPLES:BOOL=ON -DKokkosKernels_ENABLE_TPL_CUSPARSE=ON -DKokkosKernels_ENABLE_TPL_CUBLAS=ON -DKokkosKernels_ENABLE_TPL_BLAS=ON -DLAPACK_LIBRARY_DIRS=/home/projects/ppc64le-pwr8/netlib/3.8.0/gcc/7.2.0/lib -DBLAS_LIBRARY_DIRS=/home/projects/ppc64le-pwr8/netlib/3.8.0/gcc/7.2.0/lib -DLAPACK_LIBRARIES=lapack -DBLAS_LIBRARIES=blas -DCMAKE_EXE_LINKER_FLAGS=-lgfortran -lm -DCMAKE_BUILD_TYPE=DEBUG $KOKKOSKERNELS_PATH
```

Running the `KokkosKernels_graph_cuda` unit test results in this loop never terminating
https://github.com/kokkos/kokkos-kernels/blob/efd1b16f0e659c1e5feb563e45fd5d4355ff12a0/src/graph/impl/KokkosGraph_Distance1Color_impl.hpp#L2340 

Adding @brian-kelley @crtrott @srajama1

**Comment by ndellingwood** [comment]
I put in PR kokkos/kokkos#3286 and confirmed the changes there resolve the failing graph algorithms tests.

**Comment by ndellingwood** [comment]
Closing this issue, resolved by merge of kokkos/kokkos#3286
